swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/TunnelArchiveByteStream.swift

```diff
@@ -0,0 +1,146 @@
+/// another IPC call to wake it up. The buffer is "flushed" either when the data is fully consumed (i.e. `close()` is
+/// called) or when the buffer size limit is hit. Only then will the app's completionHandler callback be called.
+public class TunnelArchiveByteStream: ArchiveByteStreamProtocol {
+  struct Chunk: Codable {
+    var done: Bool = false
+    var data = Data()
+  }
+
+  let chunkSize = 1024 * 1024 // 1 MB
+  let encoder = PropertyListEncoder()
+  var chunk: Chunk
+  var chunkHandler: ((Data?) -> Void)?
+  var logger: Log
+  var completionHandler: () -> Void
+
+  public init(
+    logger: Log,
+    chunkHandler: @escaping (Data?) -> Void,
+    completionHandler: @escaping () -> Void
+  ) {
+    self.completionHandler = completionHandler
+    self.logger = logger
+    self.chunk = Chunk()
+    self.chunkHandler = chunkHandler
+  }
+
+  public func read(into buffer: UnsafeMutableRawBufferPointer) -> Int {
+    // Not implemented
+    return 0
+  }
+
+  public func read(into buffer: UnsafeMutableRawBufferPointer, atOffset: Int64) -> Int {
+    // Not implemented
+    return 0
+  }
+
+  public func write(from buffer: UnsafeRawBufferPointer) -> Int {
+    guard buffer.count <= chunkSize else {
+      self.logger.error("\(#function): buffer size \(buffer.count) bytes is larger than chunk size \(chunkSize)")
+
+      return 0
+    }
+
+    if self.chunk.data.count + buffer.count >= chunkSize {
+      flush()
+    }
+
+    guard let baseAddress = buffer.baseAddress else {
+      self.logger.error("\(#function): buffer has no base address!")
+
+      return 0
+    }
+
+    let data = Data(bytes: baseAddress, count: buffer.count)
+    self.chunk.data.append(data)
+
+    self.logger.debug("appended \(data.count) bytes")
+
+    return buffer.count
+  }
+
+  public func write(from buffer: UnsafeRawBufferPointer, atOffset: Int64) -> Int {
+    // Not implemented; unused when writing to a memory buffer
+    return 0
+  }
+
+  public func seek(toOffset: Int64, relativeTo: FileDescriptor.SeekOrigin) -> Int64 {
+    // Not implemented; unused when writing to a memory buffer
+    return 0
+  }
+
+  // It's unknown when the system calls this, but we implement it just in case.
+  // Was never called during development / testing of this class.
+  public func cancel() {
+    // App will report this as an error
+    self.chunkHandler?(nil)
+
+    self.completionHandler()
+  }
+
+  // System calls this when compression is complete
+  public func close() {
+    self.chunk.done = true
+    flush()
+
+    self.completionHandler()
+  }
+
+  // App has woken us up to receive another chunk
+  public func ready(_ chunkHandler: @escaping ((Data?) -> Void)) {
+    self.chunkHandler = chunkHandler
+  }
+
+  // Queue next chunk for app to consume
+  private func flush() {
+    do {
+      while self.chunkHandler == nil {
+        // Wait until we're called again
+        Thread.sleep(forTimeInterval: 0.01)
+      }
```

Comment:
So how does this work? Is the pointer to `TunnelArchiveByteStream` shared across multiple threads? This function is essentially busy-looping until one of its variables is set again. To make this work, two independent threads need to concurrently access this data structure. Is this access synchronised by default in Swift? Is this thread-safe?

Reply:
So not quite multiple threads - Swift's concurrency model treats `Tasks` as lightweight concurrency primitives (like an eventloop or threadpool) and therefore very little manual synchronization is necessary.

Only one instance of `TunnelArchiveByteStream` is kept on the `PacketTunnelProvider` by storing the instance in the PacketTunnelProvider's `.inProgress(FilePath, TunnelArchiveByteStream, LogCompressor)` logExportState instance variable.

The PacketTunnelProvider is only ever instantiated once by the system when the System Extension service loads us, so we won't expect for this routine to be called in multiple places.

Swift takes care of variable / state synchronization into the `Task {}` closure just like it would work for any other closure.

The thing we need to be careful about is not creating more than one `Task {}` each time the app calls us -- this is handled by the `logExportState` instance variable in PacketTunnelProvider which acts like a guard against two "threads" from entering the `inProgress` export state at the same time for the same PacketTunnelProvider instance.